lib/Common/Memory/SmallFinalizableHeapBlock.cpp

```diff
@@ -103,6 +141,146 @@ SmallFinalizableHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsi
 #endif
 }
 
+#ifdef RECYCLER_VISITED_HOST
+template <class TBlockAttributes>
+void
+SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsigned char attributes)
+{
+    // Don't call __super, since that has behavior we don't want (it asserts that FinalizeBit is set
+    // but recycler visited block allows traced only objects; it also unconditionally bumps the heap info
+    // live/new finalizable object counts which will become unbalance if FinalizeBit is not set).
+    // We do want the grandparent class behavior though, which actually sets the ObjectInfo bits.
+    SmallFinalizableHeapBlockT<TBlockAttributes>::Base::SetAttributes(address, attributes);
+
+#ifdef RECYCLER_FINALIZE_CHECK
+    if (attributes & FinalizeBit)
+    {
+        finalizeCount++;
+        HeapInfo * heapInfo = this->heapBucket->heapInfo;
+        heapInfo->liveFinalizableObjectCount++;
+        heapInfo->newFinalizableObjectCount++;
+    }
+#endif
+}
+
+template <class TBlockAttributes>
+template <bool doSpecialMark>
+_NOINLINE
+void SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::ProcessMarkedObject(void* objectAddress, MarkContext * markContext)
+{
+    unsigned char * attributes = nullptr;
+    if (!this->TryGetAddressOfAttributes(objectAddress, &attributes))
+    {
+        return;
+    }
+
+    if (!this->template UpdateAttributesOfMarkedObjects<doSpecialMark>(markContext, objectAddress, this->objectSize, *attributes,
+        [&](unsigned char _attributes) { *attributes = _attributes; }))
+    {
+        // Couldn't mark children- bail out and come back later
+        this->SetNeedOOMRescan(markContext->GetRecycler());
+    }
+}
+
+template <class TBlockAttributes>
+template <bool doSpecialMark, typename Fn>
+bool SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::UpdateAttributesOfMarkedObjects(MarkContext * markContext, void * objectAddress, size_t objectSize, unsigned char attributes, Fn fn)
+{
+    bool noOOMDuringMark = true;
+
+    if (attributes & TrackBit)
```

Comment:
I don't see any allocation variant in Recycler.h where we can allocated visited host objects with a track bit so when would this condition be true? Wondering if the track/newtrack code can be just removed from this method and treated as assert/failfast?

Reply:
See comment above on traced objects currently using the TrackBit.